cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,502 @@
+                cudaStream_t stream, bool setZero = true)
+    : m_num_batches(num_batches),
+      m_allocator(allocator),
+      m_cublasHandle(cublasHandle),
+      m_stream(stream) {
+    m_shape = std::make_pair(m, n);
+
+    // Allocate memory
+    auto memory =
+      BMM_Allocate(m_shape, num_batches, setZero, allocator, stream);
+
+    /* Take these references to extract them from member-storage for the
+     * lambda below. There are better C++14 ways to do this, but I'll keep
+     * it C++11 for now. */
+    auto& shape = m_shape;
+
+    /* Note: we create this "free" function with explicit copies to ensure that
+     * the deallocate function gets called with the correct values. */
+    auto f1 = [allocator, num_batches, shape, stream](double* A) {
+      allocator->deallocate(
+        A, num_batches * shape.first * shape.second * sizeof(double), stream);
+    };
+
+    auto f2 = [allocator, num_batches, stream](double** A) {
+      allocator->deallocate(A, sizeof(double*) * num_batches, stream);
+    };
+
+    // When this shared pointer count goes to 0, `f` is called to deallocate the memory
+    m_A_dense = std::shared_ptr<double>(memory.first, f1);
+    m_A_batches = std::shared_ptr<double*>(memory.second, f2);
+  }
+
+  //! Return batches
+  size_t batches() const { return m_num_batches; }
+
+  //! Return cublas handle
+  cublasHandle_t cublasHandle() const { return m_cublasHandle; }
+
+  //! Return allocator
+  std::shared_ptr<deviceAllocator> allocator() const { return m_allocator; }
+
+  //! Return stream
+  cudaStream_t stream() const { return m_stream; }
+
+  //! Return shape
+  const std::pair<int, int>& shape() const { return m_shape; }
+
+  //! Return pointer array
+  double** data() const { return m_A_batches.get(); }
+
+  //! Return pointer to the underlying memory
+  double* raw_data() const { return m_A_dense.get(); }
+
+  /**
+   * @brief Return pointer to the data of a specific matrix
+   * 
+   * @param[in]  id  id of the matrix
+   * @return         A pointer to the raw data of the matrix
+   */
+  double* operator[](int id) const {
+    return &(m_A_dense.get()[id * m_shape.first * m_shape.second]);
+  }
+
+  //! Stack the matrix by columns creating a long vector
+  BatchedMatrix vec() const {
+    int m = m_shape.first;
+    int n = m_shape.second;
+    int r = m * n;
+    BatchedMatrix toVec(r, 1, m_num_batches, m_cublasHandle, m_allocator,
+                        m_stream);
+    cudaMemcpyAsync(toVec[0], m_A_dense.get(),
+                    m_num_batches * r * sizeof(double),
+                    cudaMemcpyDeviceToDevice, m_stream);
+    return toVec;
+  }
+
+  /**
+   * @brief Create a matrix from a long vector.
+   * 
+   * @param[in]  m  Number of desired rows
+   * @param[in]  n  Number of desired columns
+   * @return        A batched matrix
+   */
+  BatchedMatrix mat(int m, int n) const {
+    const int r = m_shape.first * m_shape.second;
+    if (r != m * n)
```

Comment:
Minor knitpick: Please use our `ASSERT` macro for such checks and throwing exceptions. It should be defined in `cpp/include/cuml/common/utils.hpp`